lib/tapioca/helpers/rbi_helper.rb

```diff
@@ -93,15 +93,32 @@ def self.serialize_type_variable(type, variance, fixed, upper, lower)
 
     sig { params(name: String).returns(T::Boolean) }
     def valid_method_name?(name)
-      name == "==" || !(
-        name.to_sym.inspect.start_with?(':"', ":@", ":$") ||
-        name.delete_suffix("=").to_sym.inspect.start_with?(':"', ":@", ":$")
-      )
+      # try to parse a method definition with this name
+      iseq = RubyVM::InstructionSequence.compile("def #{name}; end", nil, nil, 0, false)
+      # pull out the first operation in the instruction sequence and its first argument
+      op, arg, _data = iseq.to_a.dig(-1, 0)
+      # make sure that the operation is a method definition and the method that was
+      # defined has the expected name, for example, for `def !foo; end` we don't get
+      # a syntax error but instead get a method defined as `"foo"`
+      op == :definemethod && arg == name.to_sym
+    rescue SyntaxError
+      false
     end
 
     sig { params(name: String).returns(T::Boolean) }
     def valid_parameter_name?(name)
-      /^([[:lower:]]|_|[^[[:ascii:]]])([[:alnum:]]|_|[^[[:ascii:]]])*$/.match?(name)
+      sentinel_method_name = :sentinel_method_name
```

Comment:
Interesting!

It's definitely better than the symbol hack and, as you said, at least we don't have to bother with it anymore since now it's handled by the Ruby VM.